### ISSUE TITLE
How can I fully control randomness for reproducible training??  

### ISSUE THREAD

**Issue by jellyheadandrew** [PaddlePaddle/Paddle#49058]
### 请提出你的问题 Please ask your question

Are there any ways to fully control randomness to reproduce training results??

I am using GPUs for training, and want to reproduce the training results using the same seed. I tried seeding randomness as below:

`import paddle; import paddle.fluid as fluid; import numpy as np; import random`
`paddle.seed(args.seed)`
`np.random.seed(args.seed)`
`random.seed(args.seed)`
`fluid.default_startup_program().random_seed = args.seed`
`fluid.default_main_program().random_seed = args.seed`

but still get different results when I use the same seed for training. Are there any other random seeds needed to be fixed?

**Comment by an1018** [comment]
please check：
1.  whether data is shuffled，if yes, make shuffle=False
2. randomcrop exists in data processing
3. There are dropout, random input in the network, etc
4. Optimizer impact（Adam etc）

**Comment by jellyheadandrew** [comment]
Yes I use randomcrop for augmentations in training, and maybe dropout too, but I want to control that kind of randomness too. Are there any ways?

**Comment by an1018** [comment]
The use of randomcrop results in different data training each time, and the results will naturally be different. You can try fix parameters to get the same data

**Comment by jellyheadandrew** [comment]
Yeah randomcrop procedure yields different results when randomness when that part is not controlled. I don't want to remove  or change parameters (such as randomcrop) in my training procedure because of that reason, what I want to do is **"control randomness in randomcrop part by seeding"** too.

As far as I know, analogous to torch library, you can 100% control randomness and yield same training results when you set these 7 lines of code:
`torch.manual_seed(random_seed)`
`torch.cuda.manual_seed(random_seed)`
`torch.cuda.manual_seed_all(random_seed) # if use multi-GPU`
`torch.backends.cudnn.deterministic = True`
`torch.backends.cudnn.benchmark = False`
`np.random.seed(random_seed)`
`random.seed(random_seed)`

EVEN IF there are any random augmentation procedures or random shuffling in dataset, or any dropout in models.


So I was asking if you can **set seed** for that _random augmentations_ or _dropout_ or such.
